### PR TITLE
Managed DNS support

### PR DESCRIPTION
**Pull request by csrwng**
Vendors in miekg/dns
Adds support for dns zone availability check.
Automatically creates managed DNS zone for clusterdeployment
when cluster deployment requests managed DNS.

Has dependency on https://github.com/openshift/external-dns/pull/4

### CODE REVIEW

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 1]
```diff
@@ -830,6 +854,61 @@ func (r *ReconcileClusterDeployment) setupClusterInstallServiceAccount(namespace
...
+	dnsZone := &hivev1.DNSZone{}
+	zoneName := types.NamespacedName{Namespace: cd.Namespace, Name: dnsZoneName(cd.Name)}
+	logger := cdLog.WithField("zone", zoneName.String())
+
+	err := r.Get(context.TODO(), zoneName, dnsZone)
+	if err != nil && !errors.IsNotFound(err) {
+		logger.WithError(err).Error("failed to fetch DNS zone")
+		return false, err
+	}
+	if err != nil {
```
Comment (twiest) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go]: Consider changing to:

* First check if `err == nil` and do the success case and return.
* Next check `errors.IsNotFound(err)` and do the not found case
* Last do the general error case.

File: pkg/controller/clusterdeployment/clusterdeployment_controller.go [hunk 2]
```diff
@@ -830,6 +854,61 @@ func (r *ReconcileClusterDeployment) setupClusterInstallServiceAccount(namespace
...
+		dnsZone = &hivev1.DNSZone{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      dnsZoneName(cd.Name),
+				Namespace: cd.Namespace,
+			},
+			Spec: hivev1.DNSZoneSpec{
+				Zone:               cd.Spec.BaseDomain,
```
Comment (twiest) [inline pkg/controller/clusterdeployment/clusterdeployment_controller.go]: Consider moving the `errors.IsNotFound(err)` case to a `create()` function.

File: pkg/controller/dnszone/zonereconciler.go [hunk 3]
```diff
@@ -107,30 +130,118 @@ func (zr *ZoneReconciler) Reconcile() error {
...
+
+	if zr.dnsZone.Spec.LinkToParentDomain {
+		err = zr.syncParentDomainLink(nameServers)
+		if err != nil {
+			zr.logger.WithError(err).Error("failed syncing parent domain link")
+			return reconcile.Result{}, err
+		}
+	}
+
+	isZoneAvailable, err := zr.soaLookup(zr.dnsZone.Spec.Zone, zr.logger)
```
Comment (twiest) [inline pkg/controller/dnszone/zonereconciler.go]: Consider renaming to `isZoneSOAAvailable` as we already have a `isZoneAvailable` in the status.

File: pkg/controller/dnszone/zonereconciler.go [hunk 4]
```diff
@@ -495,6 +624,51 @@ func (zr *ZoneReconciler) addRateLimitingStatusEntries() {
 	zr.dnsZone.Status.LastSyncTimestamp = &tmpTime
 }
 
+func parentLinkRecordName(dnsZoneName string) string {
+	return fmt.Sprintf("%s-ns", dnsZoneName)
+}
+
+func lookupSOARecord(zone string, logger log.FieldLogger) (bool, error) {
+	// TODO: determine if there's a better way to obtain resolver endpoints
+	clientConfig, _ := dns.ClientConfigFromFile("/etc/resolv.conf")
```
Comment (twiest) [inline pkg/controller/dnszone/zonereconciler.go]: consider moving this magic string to a constant.

File: pkg/controller/dnszone/zonereconciler.go [hunk 5]
```diff
@@ -495,6 +624,51 @@ func (zr *ZoneReconciler) addRateLimitingStatusEntries() {
...
+		}
+		log.WithField("server", s).Debugf("SOA query duration: %v", rtt)
+		if len(in.Answer) > 0 {
+			for _, rr := range in.Answer {
+				soa, ok := rr.(*dns.SOA)
+				if !ok {
+					logger.Debugf("Record returned is not an SOA record: %#v", rr)
+					continue
+				}
+				if soa.Hdr.Name != (zone + ".") {
```
Comment (twiest) [inline pkg/controller/dnszone/zonereconciler.go]: Consider checking if zone already has a trailing period and don't add one if it does.

File: pkg/operator/hive/externaldns.go [hunk 6]
```diff
@@ -54,7 +54,12 @@ func (r *ReconcileHiveConfig) deployExternalDNS(hLog log.FieldLogger, h *resourc
 	deployment := resourceread.ReadDeploymentV1OrDie(asset)
 
 	// Make AWS-specific changes to external-dns deployment
-	deployment.Spec.Template.Spec.Containers[0].Args = append(deployment.Spec.Template.Spec.Containers[0].Args, "--provider=aws")
+	args := deployment.Spec.Template.Spec.Containers[0].Args
+	args = append(args, "--provider=aws")
+	for _, domain := range instance.Spec.ManagedDomains {
+		args = append(args, fmt.Sprintf("--domain-filter=%s", domain))
```
Comment (twiest) [inline pkg/operator/hive/externaldns.go]: Verify that external-dns allows setting this parameter multiple times (that it's not "last one wins").

**Comment by csrwng** [comment]
@twiest comments addressed

File: pkg/operator/hive/externaldns.go (hunk 6 above)
Comment (csrwng) [reply pkg/operator/hive/externaldns.go]: I verified that this behaves as expected. Multiple arguments result in a union.

**Pull request merged**